lib/flat-compat.js

```diff
@@ -0,0 +1,298 @@
+    const keysToCopy = ["settings", "rules", "processor"];
+    const languageOptionsKeysToCopy = ["globals", "parser", "parserOptions"];
+    const linterOptionsKeysToCopy = ["noInlineConfig", "reportUnusedDisableDirectives"];
+    const pluginEnvironments = new Map();
+
+    // check for special settings for eslint:all and eslint:recommended:
+    if (eslintrcConfig.settings) {
+        if (eslintrcConfig.settings["eslint:all"] === true) {
+            return ["eslint:all"];
+        }
+
+        if (eslintrcConfig.settings["eslint:recommended"] === true) {
+            return ["eslint:recommended"];
+        }
+    }
+
+    // copy over simple translations
+    for (const key of keysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            flatConfig[key] = eslintrcConfig[key];
+        }
+    }
+
+    // copy over languageOptions
+    for (const key of languageOptionsKeysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+
+            // create the languageOptions key in the flat config
+            flatConfig.languageOptions = languageOptions;
+
+            if (key === "parser") {
+                debug(`Resolving parser '${languageOptions[key]}' relative to ${resolveConfigRelativeTo}`);
+                languageOptions[key] = eslintrcConfig[key].definition;
+                continue;
+            }
+
+            // clone any object values that are in the eslintrc config
+            if (eslintrcConfig[key] && typeof eslintrcConfig[key] === "object") {
+                languageOptions[key] = {
+                    ...eslintrcConfig[key]
+                };
+            } else {
+                languageOptions[key] = eslintrcConfig[key];
+            }
+        }
+    }
+
+    // copy over linterOptions
+    for (const key of linterOptionsKeysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            flatConfig.linterOptions = linterOptions;
+            linterOptions[key] = eslintrcConfig[key];
+        }
+    }
+
+    // move ecmaVersion a level up
+    if (languageOptions.parserOptions) {
+
+        if ("ecmaVersion" in languageOptions.parserOptions) {
+            languageOptions.ecmaVersion = languageOptions.parserOptions.ecmaVersion;
+            delete languageOptions.parserOptions.ecmaVersion;
+        }
+
+        if ("sourceType" in languageOptions.parserOptions) {
+            languageOptions.sourceType = languageOptions.parserOptions.sourceType;
+            delete languageOptions.parserOptions.sourceType;
+        }
+
+        // check to see if we even need parserOptions anymore and remove it if not
+        if (Object.keys(languageOptions.parserOptions).length === 0) {
+            delete languageOptions.parserOptions;
+        }
+    }
+
+    if (eslintrcConfig.ignorePattern) {
+        const ignore = IgnorePattern.createIgnore([eslintrcConfig.ignorePattern]);
```

Comment:
Hmm, this is really complicated. I’ll see what I can do.

Reply:
That concept is actually simple; concatenating all `ignorePatterns` then adding it to the [ignore](https://www.npmjs.com/package/ignore) package.

(As a result, end-users can override the `ignorePatterns` setting of shareable configs by negative patterns. In the config system, overriding shareable config is an essential requirement.)

For the compat layer, we can do the following process one-time for each config array (rather than each config array element):

```js
configs.unshift({
    // If ESLint should ignore the file, return `true`.
    ignores(filePath) {
        // Compute the final config for this file.
        // This filters config array elements by `files`/`excludedFiles` then merges the elements.
        // https://github.com/eslint/eslintrc/blob/b86d0c24b6fc3e941913469355ba77fccb8a47bc/lib/config-array/config-array.js#L475-L485
        const finalConfig = eslintrcConfigs.extractConfig(filePath)
        // Test the `ignorePattern` properties of the final config.
        return finalConfig.ignores(filePath)
    }
})
```

Reply:
Ah okay, let me plug that in and see what happens.

Reply:
Fixed the test. :tada: